### PR TITLE
RFC: Exception-safe event handlers

### PR DESCRIPTION
**Pull request by dherman**
This RFC modifies the design of [RFC 25](https://github.com/neon-bindings/rfcs/blob/master/text/0025-event-handler.md) with a solution to handling JavaScript exceptions that might get thrown during the process of launching the event handler.

[Rendered RFC](https://github.com/neon-bindings/rfcs/blob/7cfef089aae2a6527d1032803a8993a0fe2dec63/text/0000-exception-safe-event-handlers.md)

### CODE REVIEW

File: text/0000-exception-safe-event-handlers.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+[drawbacks]: #drawbacks
+
+The type signatures are perhaps a bit more complex. But we should lean on the API docs and examples as the way to explain the API, not the type signatures.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+Currently the `neon::event` module is protected by a feature flag, and can't be made generally available until we resolve the problem that it has no way to handle JavaScript exceptions.
+
+We should separately propose a `try_catch()` API for wrapping computations that might throw JavaScript exceptions in a closure and converting the result of the computation in a `Result`. We could leave the `neon::event` API as-is and just tell people to use that. But the high-level API would be less ergonomic, since you'd commonly have to wrap everything in a `try_catch`. Also it seems better to decouple the two APIs and release `neon::event` without having to design and implement a complete `try_catch` solution.
```
Comment (kjvalencik) [inline text/0000-exception-safe-event-handlers.md:159]: I think there would be value in implementing this API _using_ `try_catch` instead of in C++, but, that is orthogonal to this RFC.

File: text/0000-exception-safe-event-handlers.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    pub fn new<'a, C: Context<'a>, T: Value>(cx: &C, this: Handle<T>, callback: Handle<JsFunction>) -> Self;
+
+    pub fn schedule<T, F>(&self, arg_cb: F)
+        where T: Value,
+              F: for<'a> FnOnce(&mut EventContext<'a>) -> JsResult<'a, T>,
+              F: Send + 'static;
+
+    pub fn schedule_with<F>(&self, arg_cb: F)
+        where F: FnOnce(&mut EventContext, Handle<JsValue>, Handle<JsFunction>) -> NeonResult<()>,
```
Comment (kjvalencik) [inline text/0000-exception-safe-event-handlers.md:143]: Can we make `JsValue` generic so that users don't need to `upcast`?

File: text/0000-exception-safe-event-handlers.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+// Main thread:
+let this = cx.undefined();
+let handler = EventHandler::new(cx, this, f);
+
+// ...
+
+// Another Rust thread:
+handler.schedule_with(move |cx, this, f| {
+    let args = vec![ /* ... */ ];
+    f.call(cx, this, args).unwrap();
```
Comment (kjvalencik) [inline text/0000-exception-safe-event-handlers.md:28]: This usage doesn't appear to match the API. Should this changed to...

```rust
handler.schedule_with(move |cx, this, f| {
    let args = vec![ /* ... */ ];
    f.call(cx, this, args)?;

    Ok(())
});
```

Comment (dherman) [reply text/0000-exception-safe-event-handlers.md:28]: This is describing the API as it appears in RFC 25, but that's not very clear. Let me tighten this section up a bit.

File: text/0000-exception-safe-event-handlers.md (hunk 2 above)
Comment (dherman) [reply text/0000-exception-safe-event-handlers.md:143]: The user takes the `JsValue` as an argument, so they don't upcast it. They do have to downcast it if they know it's a particular type but that's not tracked statically. To track it statically we'd need to make `EventHandler` generic in the type of the `this`-binding, maybe something like:

```rust
struct EventHandler<This=JsObject> {
    pub fn new<'a, C: Context<'a>>(cx: &C, this: Handle<This>, callback: Handle<JsFunction<This>>) -> Self;

    pub fn schedule<T, F>(&self, arg_cb: F)
        where T: Value,
              F: for<'a> FnOnce(&mut EventContext<'a>) -> JsResult<'a, T>,
              F: Send + 'static;

    pub fn schedule_with<F>(&self, arg_cb: F)
        where F: FnOnce(&mut EventContext, Handle<This>, Handle<JsFunction<This>>) -> NeonResult<()>,
              F: Send + 'static;

}
```

Comment (dherman) [reply text/0000-exception-safe-event-handlers.md:143]: Oh, you probably meant for `schedule()` rather than `schedule_with()`!

Comment (dherman) [reply text/0000-exception-safe-event-handlers.md:143]: Wait, I'm still confused 😂  It looks like `schedule()` already _is_ generic and didn't need the upcast in the example!

File: text/0000-exception-safe-event-handlers.md (hunk 1 above)
Comment (dherman) [reply text/0000-exception-safe-event-handlers.md:159]: I'll add a comment about this to this section.

**Comment by dherman** [comment]
This RFC is entering final comment period! If there's no more significant discussion in the next 7 days, it will be accepted and merged!
